java/arcs/core/analysis/ClaimDeduction.kt

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.analysis
+
+import arcs.core.data.AccessPath
+import arcs.core.data.expression.Expression
+
+internal typealias Path = List<AccessPath.Selector.Field>
+internal typealias Paths = List<Path>
+internal typealias ClaimDerivations = Map<Path, Set<Path>>
+
+/** A visitor that accumulates field [Paths] from a Paxel [Expression]. */
+class ExpressionPathAccumulator : Expression.Visitor<Paths> {
+
+    override fun <L, R, T> visit(expr: Expression.BinaryExpression<L, R, T>): Paths {
+        return expr.left.accept(this) + expr.right.accept(this)
+    }
+
+    override fun <T> visit(expr: Expression.FieldExpression<T>): Paths {
+        if (expr.qualifier == null) {
+            return listOf(listOf(AccessPath.Selector.Field(expr.field)))
+        }
+        return listOf(
+            expr.qualifier!!.accept(this).first() + listOf(AccessPath.Selector.Field(expr.field))
+        )
+    }
+
+    override fun visit(expr: Expression.NewExpression): Paths =
+        expr.fields.map { it.second.accept(this) }.reduce { acc, set -> acc + set }
+
+    override fun <E, T> visit(expr: Expression.UnaryExpression<E, T>): Paths {
+        TODO("Not yet implemented")
+    }
+
+    override fun <T> visit(expr: Expression.QueryParameterExpression<T>): Paths {
+        TODO("Not yet implemented")
+    }
+
+    override fun visit(expr: Expression.NumberLiteralExpression): Paths {
+        TODO("Not yet implemented")
+    }
+
+    override fun visit(expr: Expression.TextLiteralExpression): Paths {
+        TODO("Not yet implemented")
+    }
+
+    override fun visit(expr: Expression.BooleanLiteralExpression): Paths {
+        TODO("Not yet implemented")
+    }
+
+    override fun visit(expr: Expression.NullLiteralExpression): Paths {
+        TODO("Not yet implemented")
+    }
+
+    override fun visit(expr: Expression.FromExpression): Paths {
+        TODO("Not yet implemented")
+    }
+
+    override fun visit(expr: Expression.WhereExpression): Paths {
+        TODO("Not yet implemented")
+    }
+
+    override fun <T> visit(expr: Expression.SelectExpression<T>): Paths {
+        TODO("Not yet implemented")
+    }
+
+    override fun visit(expr: Expression.LetExpression): Paths {
+        TODO("Not yet implemented")
+    }
+
+    override fun <T> visit(expr: Expression.FunctionExpression<T>): Paths {
+        TODO("Not yet implemented")
+    }
+}
+
+/** A visitor that accumulates [ClaimDerivations] from a Paxel [Expression]. */
+class ExpressionClaimDeducer : Expression.Visitor<ClaimDerivations> {
```

Comment:
I have given this some thought this morning, let me list a few complications that we need to handle:
1. Paxel expressions can return arbitrarily deep hierarchical objects, as well as primitive values.
2. We need to analyze both strict derivation as well as fields used for selection (WHERE, JOIN ... ON, ORDERBY)

We need to be quite flexible in how we analyze those expressions, some expressions that are valid but may not seem obvious:

`(new Foo {a: foo.x + foo.y, b: foo.y + foo.z}).a` - here construct a scope and looku it up straight away, we need to understand the analyzer result for both `a` and `b` identifiers and look up just a.

`new Foo {a: from b in bar where b.x > 42 select b.y + b.z}` - here we have entire paxel expression as a field on an object, we need to understand that b.x is used for selection while b.y and b.z for derivation.

If I were doing this I would structure the result of a Claim Deduction as a recursive data structure. E.g. something like this, though that's not exactly correct I think, but to get a flavour:

```
typealias Identifier = String
typealias Path = List<Identifier>

/** Result of the Path Analysis on an [Expression] */
sealed class PathsAnalysis {

  val selectionPaths: List<Path>

  /**
    * Result of the Path Analysis for an Expression that returns
    * a primitive value (e.g. Integer, String, Boolean), or a sequence
    * of primitive values.
    */
  class PrimitivePathsAnalysis : PathsAnalysis {
    val derivationPaths: List<Path>
  }

  /**
    * Result of the Path Analysis for an Expression that returns
    * a scope (e.g. Paxel Scope, Entity) or a sequence of scopes.
    */
  class ScopePathsAnalysis : PathsAnalysis {
    val derivationPaths: Map<Identifier, PathsAnalysis>
  }
}
```

Where refactoring paxel so that expressions like LET, FROM, WHERE, ORDERBY return a Scope object, which represents variables introduced into the scope, so all these variables can be described in the `derivationPaths` field of the returned `ScopeValuePathsAnalysis` for these expressions.

We could walk through some non trivial example of how this could work together, but probably better done over VC.

Reply:
I've taken an approach that combines your two general suggestions. Now, the visitor operates over a tuple. This tuple collects top-down claim relationships in one argument, and bubbles bottom-up information in the other. 

The "relationships" I'm referring to are a recursively define data structure similar to the one you propose above. One key difference is that I've added two other cases to define the types of associations, i.e. the particular type of claim relationship.

This may be a bit confusing to ready, but hopefully it's clear when reading the code + documentation of what I'm doing. I think we can further clarify / narrow in on a design over VC.